Bool mixed affinity support

mkgrgis opened the pull request:
This PR include:

* ISO:SQL `bool` support where some text caseinsensetive values treated as boolean values both in `SELECT` and `WHERE`. Mixed affinity values support borrowed from code lines about UUID.
* Multiversional separate `bool` support tests. No differences between versions noticed.
* Refactor `uuid_extension.c` to `sqlite_data_norm.c` with changing indent style to PostgreSQL. This file in future PRs will contain SQLite internal functions for mixed affinity cases: `float` text `Infinity` equal to SQLite `9e999`, maybe timestamps etc.
* Reduce `coalesce` during deparsing UUID normalization function.
* Refactor SQLite C-language UUID normalization function.
* Improve UUID transfer productivity up to 10-15% because of 2 previous changes.
* Fix some `gcc` 10.2+ warnings with PostgreSQL 16+.
* Add `uuid` multi-versional tests and exclude to separate UUID test from `type` test.
* Description about different UUID support in PostgreSQL 14- and PostgreSQL 14+ caused different `bind` call context.
* Deny `text` affinity for PostgreSQL `uuid` column because of SQLite C-language UUID normalization function always gets result with `blob` affinity.
* Improve error message about disallowed empty `text` during reading data for PostgreSQL column.

There are no UUID functionality changes, only code refactoring.

mkgrgis commented:
@t-kataym please start review of this PR by pgspider team when You'll have enough time for this.
I have no testing problems with this branch in my multiversional testing environment now.

If this PR is big for pgspider team work plan now, You can start review of small PRs when You'll have enough time. Small PRs see https://github.com/pgspider/sqlite_fdw/pull/89 or https://github.com/pgspider/sqlite_fdw/pull/89. It have got no conflicts with this PR.

t-kataym commented:
@mkgrgis Thank you for the PR. We will confirm it.

nxhai98 commented on License:
```diff
@@ -1,6 +1,8 @@
-SQLite Foreign Data Wrapper for PostgreSQL
+SQLite Foreign Data Wrapper for PostgreSQL linensed under
```
Could you revert this file. It's our template.

mkgrgis replied:
What exactly is not correct? This is not "The PostgreSQL Licence" or I have mistaken in `2024`? Does it means Toshiba Corporation doesn't upgrade this FDW in 2019, 2020, 2021, 2022 and 2023? Or any changes after 2018 doesn't protected by the license? I understand this is template, but what if this template is outdated? What do you law department think about this?

nxhai98 replied:
Our rule does not have last modified year.     
And this FDW created based on mysql_fdw, so we keep the License same as mysql_fdw include the notice:
`Copyright (c) 2011 - 2016, EnterpriseDB Corporation`

And also, mysql_fdw does not have "The PostgreSQL Licence" notice.

mkgrgis replied:
Thanks for explanation, reverted in https://github.com/pgspider/sqlite_fdw/pull/88/commits/6ff3790a8a7970f26cfb11e4d4472f24a678cb85

nxhai98 replied:
Thank you for fixing. Please help me to resolve this code.

mkgrgis replied:
I'll try to ask `mysql_fdw` authors about complete description of `mysql_fdw` license. (Note: `mysql_fdw` doesn't look like active project, still 19 PRs )

nxhai98 commented on README.md:
```diff
@@ -35,7 +35,14 @@ Features
+ 	- `bool`: `text`(1..5) and `int`.
+- Support mixed SQLite [data affinity](https://www.sqlite.org/datatype3.html) output (`INSERT`/`UPDATE`) for such dataypes as
+	- `timestamp`: `text`(default) or `int`,
+ 	- `uuid`: `text`(36) or `blob`(16)(default).
```
How about insert `bool`?

nxhai98 replied:
Thank for your explaining. I understood.    
Please help me to resolve this thread.

nxhai98 commented on README.md:
```diff
@@ -615,9 +636,10 @@ Useful links
 
 License
 -------
+This FDW linensed under PostgreSQL-like license.
```
Please revert `License` section.

mkgrgis replied:
Reverted in https://github.com/pgspider/sqlite_fdw/pull/88/commits/b78b5ed375265564fba2bb9d07d71669f89f8d32

nxhai98 replied:
Could you revert Copyright notice also?

mkgrgis replied:
Reverted in https://github.com/pgspider/sqlite_fdw/pull/88/commits/e81840af1156fc10c8a427591a3f5718bd42519c , please check again

nxhai98 replied:
Thank you. Please help me to resolve this thread.

nxhai98 commented on test.sh:
```diff
@@ -1,15 +1,15 @@
-rm -rf /tmp/*.data
-rm -rf /tmp/sqlitefdw_test*.db
-cp -a sql/init_data/*.data /tmp/
+rm -rf /tmp/sqlitefdw_test*.db;
```
Could you explain why `;` character added?

mkgrgis replied:
This have added as command separator. Sometimes during email transfer line separator in transferred text files can be changed from `\n` to `\r` or something similar. This tiny change allows to ignore this problems.

nxhai98 replied:
Thank you.   
Please help me to resolve this thread.

nxhai98 commented on sqlite_fdw.c:
```diff
@@ -3143,14 +3145,14 @@ sqliteImportForeignSchema(ImportForeignSchemaStmt *stmt,
-				
-				rc = sqlite3_step(pragma_stmt);
-				if (rc == SQLITE_DONE)
+				int			rc1 = sqlite3_step(pragma_stmt);
```
Is this fix for gcc warning?    
Currently we are using gcc4 (default of CentOS 7) and have no warning for it.   
Could you please share your env spec?

mkgrgis replied:
> Is this fix for gcc warning?

Yes. This and some nearest changes with variable names.

> Could you please share your env spec?

Yes. My output `gcc -v` shows more newer version, **`gcc 10.2.1 20210110`**. Also this warning was only during PostgreSQL 16 compile process.

```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/10/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.1 20210110
```

nxhai98 replied:
Thank you for share. Please help me resolve this thread.

nxhai98 commented on expected/12.15/extra/bool.out:
```diff
@@ -0,0 +1,251 @@
+--Testcase 29:
+INSERT INTO "type_BOOLEAN" (i, b) VALUES (26, NULL);
+--Testcase 30:
+SELECT * FROM "type_BOOLEAN";
```
Could you add test for float value?

mkgrgis replied:
> Could you add test for float value?

Yes. This is a good idea, @nxhai98 ! This TC should be with error.

nxhai98 replied:
Thank you.   
Please help me to resolve this thread.

nxhai98 commented on expected/12.15/extra/bool.out:
```diff
@@ -0,0 +1,251 @@
+(13 rows)
+
+--Testcase 40:
+SELECT * FROM "type_BOOLEAN+" WHERE NOT b;
```
Could you add more boolean operator test like AND, OR, XOR? 
Could you also add the explain query to show the push down behavior?

Note: You may need more one column to avoid Postgres optimization.

mkgrgis replied:
> Could you add more boolean operator test like AND, OR, XOR?

No problems, this will be added. I think this should be tests around of pushdowning.

mkgrgis replied:
`AND` and `OR` tests have been added for both `SELECT` and `SELECT ... WHERE` with pushdowning. Also mainly tested `UPDATE ... WHERE` and `DELETE ... WHERE`.

nxhai98 replied:
Thank you.   
Please help me to resolve this thread.

nxhai98 commented on expected/13.11/extra/bool.out:
```diff
@@ -0,0 +1,251 @@
+(13 rows)
+
+--Testcase 40:
+SELECT * FROM "type_BOOLEAN+" WHERE NOT b;
```
Could you add test for UPDATE/INSERT/DELETE for this table also?

mkgrgis replied:
This is not store table, @nxhai98 , this is a view - SQLite storage indicator for `type_BOOLEAN` data. All possible `INSERT` yet tested with a binded new values `0` and `1`, `UPDATE` and `DELETE` will be added. Thanks for idea!

mkgrgis replied:
`UPDATE` and `DELETE` tests added in https://github.com/pgspider/sqlite_fdw/pull/88/commits/2a2261e3bb97b737c260ff2471f3850b4df116e4 . Please confirm TCs, @nxhai98 .

nxhai98 replied:
Thank you.   
Please help me to resolve this thread.

nxhai98 replied:
Thank you for updating. Please help me to resolve this thread.

nxhai98 reviewed (changes requested):
@mkgrgis I completed my 1st review. Could you check these comment?

mkgrgis commented:
> @mkgrgis I completed my 1st review.

Thanks @nxhai98 ! First round is ended, let's begin the second.

nxhai98 commented on expected/13.11/extra/bool.out:
```diff
@@ -0,0 +1,3652 @@
+SELECT * FROM "type_BOOLEAN_oper";
+  i  | i1 | b1 | i2 | b2 
+-----+----+----+----+----
+   1 |  1 | t  |  1 | t
```
This table is too long to verify result after adding filter. Could you limit it to about 10 rows?

mkgrgis replied:
SQLite tests shows us full possible input variants. So, this table represents some main possible combinations (unfortunately not all!) of boolean data with `text` affinity. This is yet reduced test. Please refer 
```sql
CREATE TABLE "type_BOOLEAN_oper" AS
WITH booldata AS (
	SELECT row_number() over () i, column1 AS b
	  FROM (VALUES ('Yes'), ('YeS'), ('yes'), ('no'),('No'), ('nO'), ('off'),('oFf'),('on'),('ON'),('t'),('T'),('Y'),('y'),('F'),('f'),('n'),(0),(1),('0'),('1'), (NULL))
```
What is your opinion about this, @nxhai98 ? Something like `LIMIT 10` is possible, but in this case a calculations with some combinations will be untested. I have check this tables with `i1` and `i2` list for needed TCs. The table represents cartesian product of initial data.

mkgrgis replied:
@nxhai98 , maybe `1::bool` firsts and than `0::bool` will be helpful for simple checking?

```sql
(VALUES
('Yes'), ('YeS'), ('yes'),('on'),('ON'),('t'),('T'),('Y'),('y'),(1),,('1'),('1'),('true'),('tRuE'),
('no'),('No'), ('nO'), ('off'),('oFf'),('F'),('f'),('n'),(0),('0'),('false'),('fALsE'),
(NULL))`
```

nxhai98 replied:
I think the test for `bool literal` combinations unnecessary. For example:   
Combination `sqlite_fdw_bool('tRuE') AND sqlite_fdw_bool('false')` depends on the result of `sqlite_fdw_bool` for each bool literal and it has been tested before.    

and the main purpose of this case is bool operator pushed down.

mkgrgis replied:
Thanks, I will reduce this to `DISTINCT` without row number.

mkgrgis replied:
Fixed in https://github.com/pgspider/sqlite_fdw/pull/88/commits/6ccdf191564693acfe56375bde3d7a6265b65b21 to unique combinations only, please verify @nxhai98.

nxhai98 replied:
Thank you. please help me resolve this thread.

nxhai98 commented on sqlite_fdw.c:
```diff
@@ -3932,16 +3934,18 @@ sqlite_foreign_grouping_ok(PlannerInfo *root, RelOptInfo *grouped_rel)
 	 */
 	if (fpinfo->local_conds)
 	{
+		List	   *aggvars1 = NIL;
```
Could you make same as postgres_fdw:  
- remove definition in L3759
- add new definition in L3854
- revert the changed here.

mkgrgis replied:
In my current git file state point there is other line numbers, @nxhai98 . I have changed some near lines in https://github.com/pgspider/sqlite_fdw/pull/88/commits/08d87482359053057ff47d4418fbd265069a5a25 and have no failed multiversional tests. Please verify current state and let tell me where where is incorrect declarations now.

nxhai98 replied:
Thank you. It's same as my expectation.
Please help me to resolve this thread.

nxhai98 reviewed (commented):
I add some more comment for Warning fixing. Could you check that.

mkgrgis commented on sqlite_fdw.c:
```diff
@@ -3143,14 +3145,14 @@ sqliteImportForeignSchema(ImportForeignSchemaStmt *stmt,
 				char	   *default_val;
 				int			primary_key;
-				
-				rc = sqlite3_step(pragma_stmt);
```
After `git pull` and new PostgreSQL 16.0 source code getting I have also no warnings here. Reverted in https://github.com/pgspider/sqlite_fdw/pull/88/commits/08d87482359053057ff47d4418fbd265069a5a25.

nxhai98 replied:
Thank you. Please help me to resolve this thread.

mkgrgis commented on sqlite_fdw.c:
```diff
@@ -1114,12 +1114,14 @@ sqliteGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
 		double		rows;
 		int			width;
+		Cost		startup_cost1;
+		Cost		total_cost1;
```
Changed in https://github.com/pgspider/sqlite_fdw/pull/88/commits/08d87482359053057ff47d4418fbd265069a5a25 This will be really better names, Thanks @nxhai98 !

nxhai98 replied:
Thank you. Please help me to resolve this thread.

mkgrgis commented:
2st review round is ended, @nxhai98 ! Thanks. What will be later?

nxhai98 commented:
> 2st review round is ended, @nxhai98 ! Thanks. What will be later?

Thank you for your PR, I completed my review.   
@t-kataym will review your PR. Please wait.

mkgrgis commented:
@t-kataym , could you please tell me about possible review date?

t-kataym commented:
I'm sorry for late response.
I would like to review in the next week.

mkgrgis commented:
Thanks, @t-kataym ! It's important to have approximate PR plan for me. I hope better by ISO:SQL [datatype-affinity table](https://github.com/pgspider/sqlite_fdw/tree/master#datatypes) implementation will be merged in 5 PRs before 2024 year. Such PRs as `int` out of range, UUID fix with tests refactoring and `real` infinity presentation yet implemented with tests.

t-kataym commented on sqlite_data_norm.c:
```diff
@@ -0,0 +1,303 @@
+
+#include "sqlite3.h"
+
+int sqlite_fdw_data_norm_functs_init(sqlite3* db);
```
This line is unnecessary.

mkgrgis replied:
Are you about line72 , @t-kataym  ?
In this case with PostgreSQL 16 + gcc 0.2.1 20210110
```
sqlite_data_norm.c:287:1: warning: no previous prototype for ‘sqlite_fdw_data_norm_functs_init’ [-Wmissing-prototypes]
  287 | sqlite_fdw_data_norm_functs_init(sqlite3* db) {
      |
```
`int sqlite_fdw_data_norm_functs_init(sqlite3* db);` in not in `sqlite_fdw.h` because there is only 1 function with outer usage in the file.

mkgrgis replied:
Refactoring is possible, changed by adding `#include "sqlite_fdw.h"`.

mkgrgis replied:
Fixed in https://github.com/pgspider/sqlite_fdw/pull/88/commits/72e1b22177db37251b2cda4474eb46955d1d6f26, please check again.

t-kataym replied:
> Are you about line72 , @t-kataym ?
> Refactoring is possible, changed by adding #include "sqlite_fdw.h".

Yes. Your refactoring is as my expectation. Thank you for fixing.

t-kataym commented on sqlite_data_norm.c:
```diff
@@ -0,0 +1,303 @@
+			return;
+		}
+	}
+	if ( l == 2 )
```
I think it id better to use `else if`. Line 261, 274 and 279 are also same.

mkgrgis replied:
Up to 3-4% productivity on 3Gb test. Thanks, @t-kataym! Fixed in https://github.com/pgspider/sqlite_fdw/pull/88/commits/72e1b22177db37251b2cda4474eb46955d1d6f26.

t-kataym replied:
Thank you for fixing.

t-kataym commented on sqlite_data_norm.c:
```diff
@@ -0,0 +1,303 @@
+	
+	static const int det_flags = SQLITE_UTF8 | SQLITE_INNOCUOUS | SQLITE_DETERMINISTIC;
+
+	int rc = sqlite3_create_function(db, "sqlite_fdw_uuid_blob", 1, det_flags, 0, sqlite_fdw_data_norm_uuid, 0, 0);	
```
Please check the return code and raise an error by ereport()/elog() with an error message acquired by sqlites_errmsg() if an error occurred.

The return type of function will be `void`. So could you fix connection.c also where calling sqlite_fdw_data_norm_functs_init()?

mkgrgis replied:
>  ... could you fix connection.c also where calling sqlite_fdw_data_norm_functs_init()?

Yes. Distributed error processing is bad practice for this very simple case. Concentrated in `sqlite_data_norm.c` in https://github.com/pgspider/sqlite_fdw/commit/72e1b22177db37251b2cda4474eb46955d1d6f26. Please recheck, @t-kataym .

t-kataym replied:
Thank you for fixing.

t-kataym commented:
@mkgrgis I'm sorry for keeping you waiting. I finished my 1st review. Could you confirm my comments?

mkgrgis commented:
> Could you confirm my comments?

Yes, @t-kataym. All is confirmed and fixed in https://github.com/pgspider/sqlite_fdw/pull/88/commits/72e1b22177db37251b2cda4474eb46955d1d6f26.

mkgrgis commented:
I have no other commits, @t-kataym . It was useful `sed -i -e's/[ \t]*$//' ./*.c;` for deleting tabs or whitespaces at the ends of code lines.

t-kataym commented:
@mkgrgis , Thank you for your fix.  I finished the 2nd review and there are 2 comments. Could you fix them and rebase the code?

mkgrgis commented on sqlite_data_norm.c:
```diff
@@ -0,0 +1,322 @@
+#include "postgres.h"
+#include "sqlite_fdw.h"
+
+void error_helper(sqlite3* db, int rc);
```
Yes, this will better. Fixed in https://github.com/pgspider/sqlite_fdw/pull/88/commits/2bd1510602cccb070605043e3cec3bf0f01fe794 , @t-kataym .

t-kataym replied:
Thank you for fixing.

mkgrgis commented on sqlite_data_norm.c:
```diff
@@ -0,0 +1,322 @@
+error_helper(sqlite3* db, int rc)
+{
+	const char * err = sqlite3_errmsg(db);
+	sqlite3_close(db);
```
`sqlite3_close(db);` removed in https://github.com/pgspider/sqlite_fdw/pull/88/commits/2bdcc1be2329d0371e269d3f29c94a6d30e8baad . This tiny function avoid doubled C code.

t-kataym replied:
Thank you for fixing.

mkgrgis commented:
> @mkgrgis , Thank you for your fix. I finished the 2nd review and there are 2 comments. Could you fix them and rebase the code?

Fixed in prev. commits and rebased in https://github.com/pgspider/sqlite_fdw/pull/88/commits/899191c529cbddd0fde9a8a21f666cc9f3892b53 @t-kataym ! Thanks for review. Now I have no testing problems.

t-kataym commented:
@mkgrgis Thank you for your update.  
@nxhai98 Could you confirm changes again? If no problem, I will accept the PR.

nxhai98 commented:
Hi @mkgrgis,
 
It seems your rebase process has problem: [sql/15.3](https://github.com/mkgrgis/sqlite_fdw/tree/bool_mixed_affinity_support/sql/15.3) directory should be removed.   

Could you check again?

mkgrgis commented:
> Could you check again?

Yes, deleted, @nxhai98 . Please recheck.

nxhai98 commented:
@mkgrgis,

Thank you for updating.  

@t-kataym,
I confirmed these changed.   
and I tested in my environment and all tests passed.

mkgrgis commented:
Thanks, @nxhai98!  I have no comments before merge of this PR as squashed commit.

_All changes also applied to next PR https://github.com/pgspider/sqlite_fdw/pull/91 about `int` out of range._

t-kataym commented:
@nxhai98 , Thank you for your confirmation.  
@mkgrgis , Thank you for your implementation.  
I will merge changes into master.
